File: lib/path/warshall_floyd.py
import sys

import numba as nb
import numpy as np

input = sys.stdin.readline


@nb.njit("void(i8,i8[:,:],i8)", cache=True)
def warshall_floyd(n, d, INF):
    for k in range(n):
        for i in range(n):
            d_ik = d[i][k]
            if d_ik == INF:
                continue
            for j in range(n):
                if d[i][j] > d_ik + d[k][j]:
                    d[i][j] = d_ik + d[k][j]


def main():
    V, E = map(int, input().split())

    # NOTE: (max cost of the problem) < INF << (max value of the dtype)
    #   e.g.) 1 <= V <= 100, 1 <= cost <= 1000
    #         Then, max cost is 99 * 1000
    #         - np.iinfo(np.int64).max: 9223372036854775807
    #         - np.iinfo(np.int32).max: 2147483647
    INF = 100 * 1000

    # shortest path
    dist = np.full(shape=(V, V), fill_value=INF, dtype=np.int64)

    # 0 cost for 'loop'
    for i in range(V):
        dist[i][i] = 0

    for _ in range(E):
        s, t, d = map(int, input().split())
        s -= 1
        t -= 1
        dist[s][t] = d
        dist[t][s] = d
        # ^^^ delete for directed graph

    warshall_floyd(V, dist, INF)

    print(*dist, sep="\n")


if __name__ == "__main__":
    main()

"""
<http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_1_C&lang=ja>

Example for input
4 6
1 2 1
1 3 5
2 3 2
2 4 4
3 4 1
4 3 7

[0, 1, 3, 5]
[1, 0, 2, 4]
[3, 2, 0, 6]
[5, 4, 6, 0]
"""

File: cpp/lib/path/warshall_floyd.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
// --------------------------------------------------------
template<class T> bool chmin(T& a, const T b) { if (b < a) { a = b; return 1; } return 0; }
#define FOR(i,l,r) for (int i = (l); i < (r); ++i)
#define REP(i,n) FOR(i,0,n)
using VLL = vector<ll>;
using VVLL = vector<VLL>;
static const ll INF = (1LL << 62) - 1;  // 4611686018427387904 - 1
// --------------------------------------------------------


void warshall_floyd(int N, VVLL& G) {
    REP(k, N) {
        REP(i, N) {
            if (G[i][k] == INF) continue;
            REP(j, N) {
                if (G[k][j] == INF) continue;
                chmin(G[i][j], G[i][k] + G[k][j]);
            }
        }
    }
}


// Verify: http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_1_C
int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout << fixed << setprecision(10);

    int V, E; cin >> V >> E;
    VVLL G(V, VLL(V, INF));
    REP(i, V) G[i][i] = 0LL;

    int s, t; ll d;
    REP(_, E) {
        cin >> s >> t >> d;
        G[s][t] = d;
    }

    warshall_floyd(V, G);

    // NEGATIVE CYCLE
    REP(i, V) {
        if (G[i][i] < 0) {
            cout << "NEGATIVE CYCLE\n";
            return 0;
        }
    }

    REP(i, V) {
        REP(j, V) {
            if (j != 0) cout << " ";
            if (G[i][j] == INF) {
                cout << "INF";
            } else {
                cout << G[i][j];
            }
        }
        cout << '\n';
    }

    return 0;
}
